proto-lens/src/Data/ProtoLens/Encoding.hs

```diff
@@ -43,9 +46,41 @@ import Lens.Family2 (set, over, (^.), (&))
 -- | Decode a message from its wire format.  Returns 'Left' if the decoding
 -- fails.
 decodeMessage :: Message msg => B.ByteString -> Either String msg
-decodeMessage input =
-    parseOnly (Parse.manyTill getTaggedValue endOfInput) input
-        >>= taggedValuesToMessage
+decodeMessage input = parseOnly (parseMessage end) input
+  where
+    end = endOfInput *> pure Nothing
+
+-- | Parse a message with the given ending delimiter (which will be EndGroup in
+-- the case of a group, and end-of-input otherwise).  The ending delimiter
+-- Parser returns (Just err) if it encountered an unrecoverable error; in
+-- particular, if it found an EndGroup with an unexpected tag number.
+parseMessage :: forall msg . Message msg => Parser (Maybe String) -> Parser msg
+parseMessage end = do
+    (msg, unsetFields) <- join $ loop def requiredFields
+    if Map.null unsetFields
+        then return $ reverseRepeatedFields fields msg
+        else fail $ "Missing required fields "
+                        ++ show (map fieldDescriptorName
+                                    $ Map.elems $ unsetFields)
+  where
+    fields = fieldsByTag descriptor
+    requiredFields = Map.filter isRequired fields
+    loop :: msg -> Map.Map Tag (FieldDescriptor msg)
+            -> Parser (Parser (msg, Map.Map Tag (FieldDescriptor msg)))
+    -- Note that the following won't work:
+    -- end >>= \case {Just err -> fail err; ...}
+    -- since otherwise the below <|> will cause us to ignore an error that's
+    -- supposed to be unrecoverable.
+    loop msg unsetFields = (end >>= pure . \case
```

Comment:
Done, good suggestion.  It turned out that the code already fails implicitly if the EndGroup tag doesn't match; I added a unit test for this.